linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -3390,6 +3403,13 @@ private[optimizer] abstract class OptimizerCore(config: CommonPhaseConfig) {
               case LongLiteral(0) => nonConstant()
               case _              => constant(lhsLit, rhsLit)
             }
+          case String_charAt =>
+            (lhsLit, rhsLit) match {
+              case (StringLiteral(lhsStr), IntLiteral(rhsInt)) if rhsInt < 0 || rhsInt >= lhsStr.length() =>
```

Comment:
I realize this is an existing pattern, but I don't think it is very good: It separates validation and folding for individual operations for now good reason. If we move the checks inside the large switch of op types above, it'll be much easier to reason that things are correct (and since `nonConstant` is a mere forwarder, duplicate code is not a concern).

Reply:
Good point. I added a commit that refactors this whole thing.